suffuse/root.go

```diff
@@ -0,0 +1,86 @@
+package suffuse
+
+import (
+  "os"
+  "bazil.org/fuse"
+)
+
+/** An interface encapsulating the essential fuse operations.
+ *  Each method returns nil if inapplicable. The fuse operations
+ *  use the first non-nil return value found for a given Path.
+ */
+type Rule interface {
+  MetaData(Path)*fuse.Attr
+  FileData(Path)[]byte
+  DirData(Path)[]fuse.Dirent
+  LinkData(Path)*Path
+}
+
+/** Eventually we'll read this kind of information out of a config
+ *  file, but for now it's hardcoded here.
+ */
+var Rules = []Rule {
+  IdRule{},  // The identity fs - passes underlying paths through unchanged
+  SedRule{}, // The sed fs - parses a '#...' suffix and applies it to the prefix path
+}
+
+/** Default implementations. A Rule struct can embed BaseRule
+ *  and then not have to implement irrelevant methods. There's
+ *  no default for Metadata because a filesystem which never
+ *  returns metadata is a useless filesystem.
+ */
+type BaseRule struct { }
+func (BaseRule) FileData(path Path)[]byte       { return nil }
+func (BaseRule) DirData(path Path)[]fuse.Dirent { return nil }
+func (BaseRule) LinkData(path Path)*Path        { return nil }
+
+/** Only marker interfaces at the moment. They will carry data
+ *  when filesystems become more complex.
+ */
+type IdRule  struct { BaseRule }
+type SedRule struct { BaseRule }
```

Comment:
Yes, that's one of the comic aspects of this endeavor. Not even "one day" but like "next week" if we were sufficiently interested. It's why we should heavily prioritize shipping something stable, even if it does barely anything. Simply being able to organize our own source code is reason enough.